packages/@glimmer/runtime/lib/curried-value.ts

```diff
@@ -46,34 +45,40 @@ export class CurriedValue<T extends CurriedType = CurriedType> {
   }
 }
 
+interface ResolvedCurriedValue<T> {
+  definition: T;
+  owner: Owner;
+  resolved: boolean;
+  positional: Reference[] | undefined;
+  named: Record<string, Reference> | undefined;
+}
+
 export function resolveCurriedValue(
-  curriedValue: CurriedValue<CurriedType.Component>,
-  args: VMArgumentsImpl
-): [string | object, Owner, boolean];
+  curriedValue: CurriedValue<CurriedType.Component>
+): ResolvedCurriedValue<object | string>;
 export function resolveCurriedValue(
-  curriedValue: CurriedValue<CurriedType.Helper> | CurriedValue<CurriedType.Modifier>,
-  args: VMArgumentsImpl
-): [object, Owner, boolean];
+  curriedValue: CurriedValue<CurriedType.Helper> | CurriedValue<CurriedType.Modifier>
+): ResolvedCurriedValue<object>;
 export function resolveCurriedValue(
-  curriedValue: CurriedValue<CurriedType>,
-  args: VMArgumentsImpl
-): [string | object, Owner, boolean] {
+  curriedValue: CurriedValue<CurriedType>
+): ResolvedCurriedValue<object | string> {
   let currentWrapper = curriedValue;
-  let prependArgs: Reference[] | undefined;
+  let positional: Reference[] | undefined;
+  let named: Record<string, Reference> | undefined;
   let definition, owner, resolved;
 
   while (true) {
     let { [ARGS]: curriedArgs, [INNER]: inner } = currentWrapper;
 
     if (curriedArgs !== null) {
-      if (curriedArgs.positional.length > 0) {
-        prependArgs =
-          prependArgs === undefined
-            ? curriedArgs.positional
-            : curriedArgs.positional.concat(prependArgs);
+      let { named: curriedNamed, positional: curriedPositional } = curriedArgs;
+
+      if (curriedPositional.length > 0) {
+        positional =
+          positional === undefined ? curriedPositional : curriedPositional.concat(positional);
       }
 
-      args.named.merge(curriedArgs.named);
+      named = named === undefined ? assign({}, curriedNamed) : assign({}, curriedNamed, named);
```

Comment:
Hmm, why do we need this ternary? Object.assign ignores undefined/null values, and so does our helper

https://github.com/glimmerjs/glimmer-vm/blob/611b2e4b2fe3fe2677ff31b10c8350eb52ee8609/packages/%40glimmer/util/lib/object-utils.ts#L28